doc/man3/EVP_EncryptInit.pod

```diff
@@ -373,7 +373,11 @@ exists.
 =item EVP_EncryptUpdate()
 
 Encrypts I<inl> bytes from the buffer I<in> and writes the encrypted version to
-I<out>. This function can be called multiple times to encrypt successive blocks
+I<out>. The pointers I<out> and I<in> may point to the same location, in which
+case the encryption will be done in-place. If they point to different locations,
+the buffers must not (partially) overlap, otherwise the encryption fails.
```

Comment:
But we (apart from legacy codepaths) currently do not check partially overlapping buffers so the function won't return failure AFAIK.

Reply:
Hmmm....is it a bug that we don't check this, if we do so in the legacy path?

Reply:
IMO no, I would not bother. Or, if anywhere, I would add the check to libcrypto, not to the providers.

Reply:
It's actually an inconsistency between frontend (EVP) and backend (provider) documentation, and i added the 'must' to align the backup documentation with the frontend documentation. But the better solution is probably to fix the frontend documentation.

(1) The frontend EVP documentation effectively states "The buffers **must** not partially overlap, otherwise the function will **certainly fail** (because there is an explicit check for it)". The backend (provider) documentation only promises that the buffers **should** not overlap (but they **may** overlap?) and don't  give any advice to the provider developer about what to do when they indeed overlap.  _(bold words are my interpretations of the wording)_

(2) both documentations use **should** instead of **must** for precoditions too often, which in my opinion is a strictly weaker requirement: if you **should** do something, you **may** decide to do otherwise. And in that case, the expected behaviour is currently unspecified in the backend, contrary to the frontend.


### Frontend (EVP)

![grafik](https://github.com/openssl/openssl/assets/11467854/4bdbf9ba-448c-4d4a-9fe8-084031f8422a)

### Backend (Provider)

![grafik](https://github.com/openssl/openssl/assets/11467854/1048035a-eaa0-43a5-949b-13e949568a72)

In my opinion, the entire partial-overlap-check on every call is an unnecessary waste of processor time, if the documentation says you **must** not call the function with partially overlapping buffers. I don't know, why it was introduced in 2016,

7141ba3196 Fix the overlapping check for fragmented "Update" operations
5fc77684f1 evp/evp_enc.c: refine partial buffer overlap detection.
c3a73daf0a evp/evp_enc.c: check for partially[!] overlapping buffers in EVP_EncryptUpdate and EVP_DecryptUpdate. It is argued that in general case it's impossible to provide guarantee that partially[!] overlapping buffers can be tolerated.

but since the providers don't implement it anyway, let's just drop the requirement.

The simplest solution to the problem would be a documentation fix: "you **must** not pass partially overlapping buffer, otherwise you **may** get undefined behaviour." And the requirements and expected outcome should be identical in the frontend and backend documentation. Even though both documentations address a different audience, the pov should not change (from caller to implementer), because this only makes the documentation confusing. IMHO the documentation should be always be formulated from the viewpoint of the caller, even for the backend documentation: i.e., state the preconditions and the expected outcome from the viewpoint of the caller.

Reply:
In particular, I dislike the word "expections" in the backend documentations. It reverts the point of view, from top-down to buttom-up, from active to passive: "you as a provider developer can expect that you will be called in the following way".

Reply:
Yeah, I agree with changing this to `must not call` in the API doc and provider doc.

Reply:
As for **should** vs. **must**: it might be that a native speaker would use those words synonymously in natural language. But in technical documentation, there those two words specify different levels of strictness, no?

Reply:
After all, **should** is the subjunctive form of **shall**,  right?

Reply:
Which means that **shall** is equally strong as **must**, but **should** is definitely weaker.

Reply:
@t8m I was too busy writing, so I didn't notice that you replied in the meantime :-)